src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -1191,6 +1694,87 @@ public static void UndoTransition(ref CurrentState state)
 #endif
         }
 
+        /// <summary>
+        /// This input reader attempts to minimize overhead
+        /// by handling constraints outside of the loop:
+        /// 1. the position must be already valid for the input.
+        /// 2. the pattern must not contain \Z.
+        /// 3. to save memory, `maxChar` is a local variable set to the ordinal char for highest non-0 minterm
+        /// </summary>
+        private interface IOptimizedInputReader
+        {
+            public static abstract int GetPositionId(byte[] lookup, int maxChar, ReadOnlySpan<char> input,
+                int pos);
+        }
+
+        /// <summary>
+        /// This reader maps all characters > maxChar to 0
+        /// </summary>
+        private readonly struct OptimizedSmallInputReader : IOptimizedInputReader
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static int GetPositionId(byte[] lookup, int maxChar, ReadOnlySpan<char> input, int pos)
+            {
+                Debug.Assert(pos < input.Length, "pos < input.Length");
+                Debug.Assert(maxChar <= (lookup.Length + 1), $"maxChar = {maxChar}; lookup.Length = {lookup.Length}");
+                return input[pos] > maxChar ? 0 : lookup[input[pos]];
+            }
+        }
+
+        /// <summary>
+        /// This reader is effectively an array lookup for the all utf16 code units
+        /// </summary>
+        private readonly struct OptimizedFullInputReader : IOptimizedInputReader
```

Comment:
Is this really a case that needs optimizing? Surely the case where maxChar is not 65K is the vast majority case and we could delete this struct, delete the special-cases that use it, simplify all that code, and just always use the equivalent of the OptimizedSmallInputReader struct?

Reply:
We can do that as well, i suppose this is only necessary when squeezing out the last couple percentages